### DISCUSSION TITLE
Qute - should a tag template be executed in isolation?

### DISCUSSION THREAD

**Discussion by mkouba** [quarkusio/quarkus#22285] [Community]
Currently, a tag template can reference data from the parent context. However, it might make sense to disable this behavior and execute the tag as an _isolated_ template, i.e. without access to the context of the template that calls the tag. There is a PR https://github.com/quarkusio/quarkus/pull/22003 where the `_isolated` argument was added to control the behavior. 

However, we're not entirely sure whether the default behavior should be changed as well, i.e. execute the tag as an isolated template by default.

The main disadvantage of the current behavior is that if an argument is not overriden on the call site a value from the parent context may unexpectedly leak to the rendered tag:
```
{! myTag !}
Name: {name}
```
and
```
{! parent template !}
{#let name='Foo'}
   {#myTag name='Bar' /} -> renders Name: Bar 
   {#myTag /} -> renders Name: Foo
{/let}
```
AND AFTER THE CHANGE:
```
{! parent template !}
{#let name='Foo'}
   {#myTag name='Bar' /} -> renders Name: Bar 
   {#myTag /} -> failure or NOT_FOUND
{/let}
```

OTOH the current behavior may come in handy in some situations. Also it would be a breaking change - although it was not documented before and it would not be the first one and definitely not the last one ;-).

**Comment by gastaldi** [comment]
I like the current behavior. It seems that are a few use-cases where isolation may be useful compared to not isolated, so IMHO we're good :)

**Comment by FroMage** [comment]
I strongly think we should model tags and scoping after what the Quarkus users are most used to, which is Java and lexical scoping. It's very confusing and error-prone to access data that doesn't come from the current scope, because copying one tag from one place to another might yield different results that are hard to understand. Also it's contradictory with the idea of passing parameters to it. And last, it pretty much breaks tag parameter declaration and type-safety.

**Reply by maxandersen** [reply]
I prefer explicit over implicit here but I can also see how it can be annoying having to pass every parameter through.

How about having syntax or explicit parameter to be able to pass the current template state to the next ? Like call in the same scope but be explicit about wanting that behavior ?

If not I'm on same side as step to make state scope/access explicit.

**Reply by FroMage** [reply]
Yeah, OK, I still don't see how it can be useful, but if we had syntax _in the tag_ to specify that it wants access to stuff that's outside of it, that's where I'd put the flag, because it can already declare its parameters, it might as well declare that it's using the Eye of Sauron to see undeclared stuff.

**Reply by maxandersen** [reply]
The time when I see the ability to access the outside scope is the exact reason I add methods to a class - to access/share state without having to pass it around constantly. In that case, one views the templates as "methods" on a shared state object rather than 100% standalone templates. 

I think this is common (but not necessarily majority). Also, it doesn't help on type-safety, so I'm still more inclined to say default is isolated, and optionally at call point, you can basically pass "this" into the call. come to think of it - kind of how Python does classes :)

**Reply by mkouba** [reply]
So the user tags were not designed to be "standalone" templates alhough they can be used like this. I think I mentioned this already but the original idea was to introduce named [include](https://quarkus.io/guides/qute-reference#include_helper)s and it was documented this way:  _"User-defined tags can be **used to include** a template and optionally pass some parameters"_. And I think that in the template world it makes sense, that an included template has access to the data from the parent scope.

> Also, it doesn't help on type-safety, ...

I don't get this argument about type-safety. Right now, an author of the tag should define the [parameter declarations](https://quarkus.io/guides/qute-reference#typesafe_expressions) for all expressions that should be validated during build. The fact, that a variable name comes from a parent scope or is a user tag param, is irrelevant. Ofc, we can do better and try to add some "call site validation" like we have for `@CheckedTemplate` but we're not there yet and I'm not quite sure it's doable... need to find some time to analyze the possibilities.

> and optionally at call point, you can basically pass "this" into the call. come to think of it - kind of how Python does classes :)

:nauseated_face: :-)

**Comment by FroMage** [comment]
> The time when I see the ability to access the outside scope is the exact reason I add methods to a class

Except in this case, it's an _include_, so it's as if that method was part of many classes, all of which have different internal state. Think macros more than methods. I prefer methods. Macros are often unhygienic: https://en.wikipedia.org/wiki/Hygienic_macro

**Comment by ia3andy** [comment]
Hey folks,

Sorry to pop this subject up but:

Yesterday I had an issue which took me a while to find and fix, a tag I was using which comes from the web-bundler use `name` as variable AND my template also used `name` because obviously I didn't think about the clash in the first place.

It feels like a tag **shouldn't** use the scope data but only what is sent through params **unless explicitly configured too**.
Something like `{#tag _scoped}` when you actually want to use the parent scope.

The current behavior is hazardous, unnatural and very hard to spot or debug.

- Includes are mostly used for layouting so passing the context makes sense.
- Tags are used as "snippets" or "components" and meant to be use many times and in different places making the default behavior of using the parent scope very risky!

**Reply by mkouba** [reply]
Thanks @ia3andy for your input.

> It feels like a tag shouldn't use the scope data but only what is sent through params unless explicitly configured too.
Something like {#tag _scoped} when you actually want to use the parent scope.

So right now, you can use `{#tag _isolated}` or `{#tag _isolated=true}` to execute the tag without access to the context of the template that calls the tag.

> The current behavior is hazardous, unnatural and very hard to spot or debug.

I personally do not see it that way but if I'm in the minority then we could change the default behavior to `_isolated=true`. It would be a breaking change but I think that it would be OK if it's properly documented.

>  Includes are mostly used for layouting so passing the context makes sense. Tags are used as "snippets" or "components" and meant to be use many times and in different places making the default behavior of using the parent scope very risky!

I'd like to stress out that in Qute includes and user tags are functionally equivalent, i.e. it does not really matter if you do `{#include myTag}` or `{#myTag}`.

**Reply by ia3andy** [reply]
and would it be possible to have `include` and `tag` be the opposite for this (we keep the current behavior for includes and change it for tags)?

**Reply by mkouba** [reply]
It's definitely possible but wouldn't it be even more confusing?

**Reply by ia3andy** [reply]
- I believe includes are meant to use the parent context as we don't usually pass params to them, you include it in the current context
- tags though are more components that you want to configure depending on where you use them and should be isolated from the current context

at least this is what would feel natural to me.

**Reply by tomocesnik** [reply]
I agree with @ia3andy. It is confusing to have tags inherit the scope. Also as @mkouba mentions, there is then no practical difference between `include` and `tag`. Having 2 almost identical things to achieve the same result is bad for maintainability/clarity of code.

**Reply by FroMage** [reply]
Agreed for making tags isolated by default! :)

As for includes, I'm not exactly sure what it means. I use includes for composition, where I declare that I include the `main` template, and I pass it parameters (using a special syntax that I also find confusing but it doesn't matter). Then within the parameters that I pass (such as the body) I don't feel as if I was using parameters declared in the `main` template. I don't even think that's possible.

Or, perhaps it's the other way around and you're saying that when I include the `main` template, it inherits the current scope and it can see all my parameters? Because if that's true, then I also find this confusing and something we should prevent, especially since we can include the `main` template from lots of different places that have widely different parameters, so I'm not sure how we can type-check this.

Probably there's something I didn't understand :)

**Reply by mkouba** [reply]
Guys, I don't read any solid arguments in the comments above. It's more about feelings, expectations and different backgrounds.

That said, I'm ready to introduce a breaking change, i.e. change the default behavior for _tags_ but keep the current behavior for _includes_. Here's a DRAFT PR: https://github.com/quarkusio/quarkus/pull/35600

If we agree that this compromise makes sense then let's close this discussion and continue on the PR...

**Reply by tomocesnik** [reply]
I agree, it is definitely about feelings and expectations - like most of the things in software. :)

From my personal point of view, I think it would be nice to have some good way of defining "components" - similar to what React has (although I am not really experienced with it), but without all the additional mechanisms. The feeling is that people mostly expect this kind of behavior from Qute tags, but I might be wrong. Also lots of people using Java might have at least some experience with JSP tags (again not an expert in it): https://docs.oracle.com/cd/E19159-01/819-3669/bnalj/index.html.

So basically, it would be cool to have a way to "include" a bunch of HTML in a page, where all the parameters are explicitly passed at the call site (no implicit `it` parameter) and the context is not implicitly inherited. Supporting `insert` sections in such a construct would maybe also go against the core philosophy (unless they are evaluated in the context of the call site?).

In any case, I think it doesn't make sense to keep `include` and `tag` functionality to share the same behavior.
Therefore I would suggest to keep this discussion open and try to partially redesign how tags work - if we all agree that this would be useful.

**Reply by ia3andy** [reply]
@tomocesnik this is what we try to achieve here:
https://docs.quarkiverse.io/quarkus-web-bundler/dev/advanced-guides.html#server-side-web-components

**Reply by mkouba** [reply]
> I agree, it is definitely about feelings and expectations - like most of the things in software. :)

:+1: :smile: 
 
> 
> From my personal point of view, I think it would be nice to have some good way of defining "components" - similar to what React has (although I am not really experienced with it), but without all the additional mechanisms. The feeling is that people mostly expect this kind of behavior from Qute tags, but I might be wrong. Also lots of people using Java might have at least some experience with JSP tags (again not an expert in it): https://docs.oracle.com/cd/E19159-01/819-3669/bnalj/index.html.
> 
> So basically, it would be cool to have a way to "include" a bunch of HTML in a page, where all the parameters are explicitly passed at the call site (no implicit `it` parameter) and the context is not implicitly inherited. Supporting `insert` sections in such a construct would maybe also go against the core philosophy (unless they are evaluated in the context of the call site?).
> 

But I don't think that components should be part of the Qute core. Qute is not JSP/JSF/etc. It's a general purpose templating library. See @ia3andy's work for web-specific stuff.

> In any case, I think it doesn't make sense to keep `include` and `tag` functionality to share the same behavior. Therefore I would suggest to keep this discussion open and try to partially redesign how tags work - if we all agree that this would be useful.

For sure, we can keep this discussion open.

**Reply by tomocesnik** [reply]
Jeah "component" is a pretty overloaded word - for the sake of convincing I just wanted to point out the main benefit of "component" "things" through a few examples. Their main point is: Encapsulation of smaller pieces to a separate/independent context.
The advantage is that one can reason about such "component" solely based on its inputs and not on some implicitly inherited state, which makes it much easier to avoid bugs as encountered by @ia3andy.

Therefore I am definitely for changing the default behavior of tags to support such encapsulation.

Points to discuss:
- Should the tags be allowed to inherit the state by using a special flag - reverse behavior as what we have now?
- Should implicit unnamed parameter `it` be supported or should all parameters be named explicitly?
- Should `insert` sections be allowed and how would they get evaluated (in which context)?

**Comment by mkouba** [comment]
https://github.com/quarkusio/quarkus/pull/35600

**Reply by sonnb** [reply]
@mkouba Anyway to make the isolation disabled by default? We have lot of running templates which access parent's variables and it is not really easy to edit them all to adapt this.

**Reply by mkouba** [reply]
Unfortunately, we don't have a config knob to disable isolation for tags globally. It would be probably even more confusing.

**Reply by sonnb** [reply]
We allow users to create their own templates then call the builtin ones. If we upgrade to newer version (We are using 3.2.12 right now), we must ask users to update all their templates to either pass down the variables where they call the builtin ones or they have to set the unisolated flags. This is quite complicated :(

**Reply by FroMage** [reply]
How many variables are you talking about? Can't they be defined/accessed as global variables? Or agregated into a container type to pass them on more easily?

**Reply by mkouba** [reply]
Global variables or a [custom namespace](https://quarkus.io/guides/qute-reference#namespace_extension_methods) would make sense but it might not always be applicable..

**Reply by sonnb** [reply]
> How many variables are you talking about? Can't they be defined/accessed as global variables? Or agregated into a container type to pass them on more easily?
Not that much but it is how it was. If I can do it from beginning at this point, I would of course choose isolated.

Somehow last time I have tried to use @TemplateGlobal but it didn't work neither with nor without global namespace. Maybe I would give it a try as it might be the only solution for now without adapt existing templates.

Is there any real example on how to initialize the values for the global values? I know it will be automatically created per TemplateInstance but not sure how to set values for it before the render method is invoked.

**Reply by mkouba** [reply]
> Is there any real example on how to initialize the values for the global values? I know it will be automatically created per TemplateInstance but not sure how to set values for it before the render method is invoked.

Global variables are just static methods/fields that are lazily set as template instance data. Where "lazily" means that static method/field is invoked/read when the name is first used in the template.

**Reply by sonnb** [reply]
Thanks @mkouba. I will try to look into it again.
